core/src/main/java/org/javarosa/core/model/actions/SetValueAction.java

```diff
@@ -71,37 +71,70 @@ public void processAction(FormDef model, TreeReference contextRef) {
         EvaluationContext context = new EvaluationContext(model.getEvaluationContext(), qualifiedReference);
 
-        Object result;
-
-        if (explicitValue != null) {
-            result = explicitValue;
-        } else {
-            result = XPathFuncExpr.unpack(value.eval(model.getMainInstance(), context));
-        }
-        
         String failMessage = "Target of TreeReference " + target.toString(true) + " could not be resolved!";
-        
-        //If the reference has predicates we need to qualify it further.
+
         if(qualifiedReference.hasPredicates()) {
-            Vector<TreeReference> references = context.expandReference(qualifiedReference);
-            if(references.size() == 0) {
+            //CTS: in theory these predicates could contain logic which breaks if the qualified ref
+            //contains unbound repeats (IE: nested repeats).
+            Vector<TreeReference> references = context.expandReference(qualifiedReference, true);
+            if (references.size() == 0) {
                 throw new NullPointerException(failMessage);
-            }
-            else if(references.size() > 1) {
-                //The spec for single value binding says we should pick the first one,
-                //but that's not how we do things to prevent errors. We're going to fail
-                //the same way a nodeset does
-                throw new XPathTypeMismatchException("XPath nodeset has more than one node [" + XPathNodeset.printNodeContents(references) + "]; Actions can only target a single node reference. Refine path expression to match only one node.");
-            }
-            else {
+            } else if (references.size() > 1) {
+                //We included templates, so it's possible we got both a ref _and_ a template. Check
+                TreeReference singleFullyQualifiedRef = null;
+                for (TreeReference ref : references) {
+                    if (!ref.isTemplateRef()) {
+                        if (singleFullyQualifiedRef == null) {
+                            singleFullyQualifiedRef = ref;
+                        } else {
+                            //The spec for single value binding says we should pick the first one,
+                            //but that's not how we do things to prevent errors. We're going to fail
+                            //the same way a nodeset does
+                            throw new XPathTypeMismatchException("XPath nodeset has more than one node [" + XPathNodeset.printNodeContents(references) + "]; Actions can only target a single node reference. Refine path expression to match only one node.");
+                        }
+                    }
+                }
+
+                if (singleFullyQualifiedRef == null) {
+                    //We only have template references, which means this setvalue action isn't relevant
+                    //for these targets yet.
+                    return;
+                }
+            } else {
```

Comment:
This template checking logic is pretty self-contained and should be a helper function.